xrdp/xrdp_egfx.c

```diff
@@ -0,0 +1,1157 @@
+        " width %d height %d monitorcount %d",
+        width, height, monitor_count);
+    if (monitor_count < 16)
+    {
+        bytes = 340 - (20 + (monitor_count * 20));
+        out_uint8s(s, bytes);
+    }
+    s_mark_end(s);
+    return s;
+}
+
+/******************************************************************************/
+int
+xrdp_egfx_send_reset_graphics(struct xrdp_egfx *egfx, int width, int height,
+                              int monitor_count, struct monitor_info *mi)
+{
+    int error;
+    struct stream *s;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_send_reset_graphics:");
+    s = xrdp_egfx_reset_graphics(egfx->bulk, width, height, monitor_count, mi);
+    error = xrdp_egfx_send_s(egfx, s);
+    LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_send_reset_graphics: xrdp_egfx_send_s "
+        "error %d", error);
+    free_stream(s);
+    return error;
+}
+
+/******************************************************************************/
+/* RDPGFX_CMDID_FRAMEACKNOWLEDGE */
+static int
+xrdp_egfx_process_frame_ack(struct xrdp_egfx *egfx, struct stream *s)
+{
+    uint32_t queueDepth;
+    uint32_t intframeId;
+    uint32_t totalFramesDecoded;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_process_frame_ack:");
+    if (!s_check_rem(s, 12))
+    {
+        return 1;
+    }
+    in_uint32_le(s, queueDepth);
+    in_uint32_le(s, intframeId);
+    in_uint32_le(s, totalFramesDecoded);
+    LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_process_frame_ack: queueDepth %d"
+        " intframeId %d totalFramesDecoded %d",
+        queueDepth, intframeId, totalFramesDecoded);
+    if (egfx->frame_ack != NULL)
+    {
+        egfx->frame_ack(egfx->user, queueDepth, intframeId, totalFramesDecoded);
+    }
+    return 0;
+}
+
+/******************************************************************************/
+/* RDPGFX_CMDID_CAPSADVERTISE */
+static int
+xrdp_egfx_process_capsadvertise(struct xrdp_egfx *egfx, struct stream *s)
+{
+    int index;
+    int capsSetCount;
+    int caps_count;
+    int version;
+    int capsDataLength;
+    int flags;
+    char *holdp;
+    int *versions;
+    int *flagss;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_process_capsadvertise:");
+    if (egfx->caps_advertise == NULL)
+    {
+        return 0;
+    }
+    in_uint16_le(s, capsSetCount);
+    if ((capsSetCount < 1) || (capsSetCount > 1024))
+    {
+        return 1;
+    }
+    caps_count = 0;
+    versions = g_new(int, capsSetCount);
+    if (versions == NULL)
+    {
+        return 1;
+    }
+    flagss = g_new(int, capsSetCount);
+    if (flagss == NULL)
+    {
+        g_free(versions);
+        return 1;
+    }
+    for (index = 0; index < capsSetCount; index++)
+    {
+        if (!s_check_rem(s, 8))
+        {
+            return 1;
+        }
+        in_uint32_le(s, version);
+        in_uint32_le(s, capsDataLength);
+        if (!s_check_rem(s, capsDataLength))
+        {
+            return 1;
+        }
+        holdp = s->p;
+        if (capsDataLength == 4)
```

Comment:
You might want to mention (for others), that caps version 101 is indirectly excluded here, but that's up to you.

Reply:
Done.